echo_server.py

```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+import socket
+
+def response_ok():
+    return u"HTTP/1.1 200 OK\r\n \
+    Content-Type: text/plain\r\n\r\n".encode('utf-8')
+
+def response_error(error_code, reason):
+    return u"HTTP/1.1 {} {}\r\n \
+    Content-Type: text/plain\r\n\r\n".format(error_code, reason).encode('utf-8')
+
+def parse_request(request):
+    first_line = request.splitlines()[0]
+    first_line = first_line.split(' ')
+    method = first_line[0]
+    uri = first_line[1]
+    protocol = first_line[2]
+
+    if method != 'GET':
+        return response_error('405', 'Method Not Allowed')
+    elif protocol != 'HTTP/1.1':
+        return response_error('505', 'HTTP Version Not Supported')
+    else:
+        return uri
+
+def server():
+    server_socket = socket.socket(
+        socket.AF_INET, 
+        socket.SOCK_STREAM,
+        socket.IPPROTO_IP)
+    server_socket.bind(('127.0.0.1', 50000))
+    server_socket.listen(1)
+    bsize = 32
+    try:
+        while True:
+            message = ''
+            done = False
+            conn, addr = server_socket.accept()
+            while not done:
+                piece = conn.recv(bsize)
+                message = '{}{}'.format(message, piece)
+                if len(piece) < bsize:
+                    done = True
+            first_line = message.splitlines()[0]
+            first_line = first_line.split(' ')
+            uri = first_line[1]
+            if str(uri) == str(parse_request(message)):
+                response = "{}{}".format(response_ok(), parse_request(message))
+            else:
+                response = parse_request(message)
```

Comment:
This bit here is the result of the way that parse_request returns a string no matter what happens.  You are forced to repeat the work it does in order to compare what it produces to its expected outcome, which is just nasty.  

It's interesting how a simple choice in how to factor your code has ramifications that spiral outwards from where the code is used.